design/multiple-storage-types-support.md

```diff
@@ -0,0 +1,272 @@
+However, using an extension apiserver has some disruptive drawbacks:
+
+* etcd must also be deployed for storage of its API objects, increasing the complexity and adding another point of failure.  This can be avoided with using a CRD to store its API objects but that is awkward and exposes internal storage to the user.
+* development cost would be significant to get the deployment working reliably in supported Rook environments and to port our existing CRDs to extension types.
+* breaking change for Rook users with no clear migration path, which would be very disruptive to our current user base.
+
+#### CustomResourceDefinitions (CRDS)
+
+CRDs are what Rook is already using to extend Kubernetes.  They are a limited extension mechanism that allows the definition of custom types, but lacks the rich features of API aggregation.  For example, validation of a users CRD is only at the schema level and simple property value checks that are available via the [OpenAPI v3 schema](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaObject).
+Also, there is currently no versioning (conversion) support.
+However, CRDs are being actively supported by the community and more features are being added to CRDs going forward (e.g., a [versioning proposal](https://github.com/kubernetes/features/issues/544)).
+Finally, CRDs are not a breaking change for Rook users since they are already in use today.
+
+### Controller Options
+
+The controllers are the entities that will perform orchestration and deployment tasks to ensure the users desired state is made a reality within their cluster.
+There are a few options for deploying and hosting the controllers that will perform this work, as explained in the following sections.
+
+#### In-proc Controllers
+
+Multiple controllers could be deployed within a single process.
+For instance, Rook could run one controller per domain of expertise, e.g., ceph-controller, minio-controller, etc.
+Controllers would all watch the same custom resource types via a `SharedInformer` and respond to events via a `WorkQueue` for efficiency and reduced burden on core apiserver.
+Even though all controllers are watching, only the applicable controller responds to and handles an event.
+For example, the user runs `kubectl create cluster.yaml` to create a `cluster.rook.io` instance that has Ceph specific properties.
+All controllers will receive the created event via the `SharedInformer`, but only the Ceph controller will queue and handle it.
+We can consider only loading the controllers the user specifically asks for, perhaps via an environment variable in `rook-operator.yaml`.
+
+Note that this architecture can be used with either API aggregation or CRDs.
+
+##### PROS
+
+* Potentially easier developer burden for new storage backends since there is no need to create a new deployment to host their controller.
+* Less resource burden on K8s cluster since watchers/caches are being shared.
```

Comment:
What is the resource burden? Seems like we're talking about some small resources.